"panic: interface conversion" while trying to shape

Repro is with `zq` commit `1ac7787f`.

I was trying to verify #2198 and was hacking around before fully understanding the example, and triggered a panic.

```
$ cat example.zs 
type port=uint16;
type conn={orig_h:ip,orig_p:port,resp_h:ip,resp_p:port};
put . = shape(., {_path:string,ts:time,uid:bstring,id:conn})

$ echo '{"foo": "bar"}' | zq -z -I example.zs -
panic: interface conversion: zng.Type is *zng.TypeAlias, not *zng.TypeRecord
goroutine 68 [running]:
runtime/debug.Stack(0x46, 0x40, 0x6b)
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x9f
github.com/brimsec/zq/zqe.RecoverError(0x182e8c0, 0xc0003bc690, 0xc0003bc690, 0xc000616200)
	/Users/phil/work/zq/zqe/error.go:163 +0x26
github.com/brimsec/zq/driver.(*mux).safeGet.func1(0xc000625f70)
	/Users/phil/work/zq/driver/mux.go:49 +0x48
panic(0x182e8c0, 0xc0003bc690)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/brimsec/zq/expr.(*Shaper).orderRecordType(0xc000132360, 0xc0003bc300, 0xc0003bc300, 0xc0003bc300, 0x0, 0x0)
	/Users/phil/work/zq/expr/shaper.go:404 +0xa10
github.com/brimsec/zq/expr.(*Shaper).createShapeSpec(0xc000132360, 0xc0003bc5a0, 0xc0003bc300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/phil/work/zq/expr/shaper.go:323 +0x1ee
github.com/brimsec/zq/expr.(*Shaper).Eval(0xc000132360, 0xc0000325f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000032501, 0xc000128c60)
	/Users/phil/work/zq/expr/shaper.go:285 +0x337
github.com/brimsec/zq/proc/put.(*Proc).eval(0xc0000a62d0, 0xc0000325f0, 0x0, 0xc000128c60, 0xc000625d98, 0x15b87dc, 0xc000128080)
	/Users/phil/work/zq/proc/put/put.go:84 +0x127
github.com/brimsec/zq/proc/put.(*Proc).put(0xc0000a62d0, 0xc0000325f0, 0xc0000325f0, 0xc000010040, 0x0)
	/Users/phil/work/zq/proc/put/put.go:347 +0x50
github.com/brimsec/zq/proc/put.(*Proc).Pull(0xc0000a62d0, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/proc/put/put.go:371 +0x15f
github.com/brimsec/zq/driver.(*mux).safeGet(0xc0001282a0, 0x0, 0x0, 0x0, 0x0)
	/Users/phil/work/zq/driver/mux.go:51 +0x6c
github.com/brimsec/zq/driver.(*mux).run(0xc0001282a0)
	/Users/phil/work/zq/driver/mux.go:64 +0x32
created by github.com/brimsec/zq/driver.(*muxOutput).Pull.func1
	/Users/phil/work/zq/driver/mux.go:98 +0x5d
```

--- a/expr/shaper.go
+++ b/expr/shaper.go
@@ -342,37 +342,44 @@ func (s *Shaper) cropRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 			continue
 		}
 
+		inType := zng.AliasedType(inCol.Type)
 		specCol := spec.Columns[ind]
+		specType := zng.AliasedType(specCol.Type)
 		switch {
-		case zng.IsPrimitiveType(inCol.Type):
+		case zng.IsPrimitiveType(inType):
 			// 2. Field is non-record in input: keep (regardless of crop record-ness)
 			cols = append(cols, inCol)
-		case zng.IsRecordType(inCol.Type) && zng.IsRecordType(specCol.Type):
+		case zng.IsRecordType(inType) && zng.IsRecordType(specType):
 			// 3. Both records: recurse
-			out, err := s.cropRecordType(inCol.Type.(*zng.TypeRecord), specCol.Type.(*zng.TypeRecord))
+			out, err := s.cropRecordType(inType.(*zng.TypeRecord), specType.(*zng.TypeRecord))
 			if err != nil {
 				return nil, err
 			}
 			cols = append(cols, zng.Column{inCol.Name, out})
-		case isCollectionType(inCol.Type) && isCollectionType(specCol.Type) &&
-			zng.IsRecordType(innerType(inCol.Type)) && zng.IsRecordType(innerType(specCol.Type)):
-			// 4. array/set of records
-			inner, err := s.cropRecordType(innerType(inCol.Type).(*zng.TypeRecord), innerType(specCol.Type).(*zng.TypeRecord))
-			if err != nil {
-				return nil, err
-			}
-			var t zng.Type
-			if _, ok := inCol.Type.(*zng.TypeArray); ok {
-				t, err = s.zctx.LookupTypeArray(inner), nil
+		case isCollectionType(inType) && isCollectionType(specType):
+			if zng.IsRecordType(innerType(inType)) && zng.IsRecordType(innerType(specType)) {
+				// 4. array/set of records
+				inInner := zng.AliasedType(innerType(inType))
+				specInner := zng.AliasedType(innerType(specType))
+				inner, err := s.cropRecordType(inInner.(*zng.TypeRecord), specInner.(*zng.TypeRecord))
+				if err != nil {
+					return nil, err
+				}
+				var t zng.Type
+				if _, ok := inCol.Type.(*zng.TypeArray); ok {
+					t, err = s.zctx.LookupTypeArray(inner), nil
+				} else {
+					t, err = s.zctx.LookupTypeSet(inner), nil
+				}
+				if err != nil {
+					return nil, err
+				}
+				cols = append(cols, zng.Column{inCol.Name, t})
 			} else {
-				t, err = s.zctx.LookupTypeSet(inner), nil
+				cols = append(cols, inCol)
 			}
-			if err != nil {
-				return nil, err
-			}
-			cols = append(cols, zng.Column{inCol.Name, t})
 		default:
-			// 5. record input but non-record in crop: keep crop
+			// 5. container input but non-container in crop: keep crop
 			cols = append(cols, specCol)
 
 		}

--- a/expr/shaper.go
+++ b/expr/shaper.go
@@ -93,15 +93,15 @@ func createStepSet(in, out zng.Type) (step, error) {
 }
 
 func isCollectionType(t zng.Type) bool {
-	switch t := t.(type) {
-	case *zng.TypeAlias:
-		return isCollectionType(t.Type)
+	switch zng.AliasedType(t).(type) {
 	case *zng.TypeArray, *zng.TypeSet:
 		return true
 	}
 	return false
 }
 
+// This is similar to zng.InnerType except it handles aliases. Should
+// be unified, see #2270.
 func innerType(t zng.Type) zng.Type {
 	switch t := t.(type) {
 	case *zng.TypeAlias:
@@ -357,10 +357,10 @@ func (s *Shaper) cropRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 			}
 			cols = append(cols, zng.Column{inCol.Name, out})
 		case isCollectionType(inType) && isCollectionType(specType):
-			if zng.IsRecordType(innerType(inType)) && zng.IsRecordType(innerType(specType)) {
+			inInner := zng.AliasedType(zng.InnerType(inType))
+			specInner := zng.AliasedType(zng.InnerType(specType))
+			if zng.IsRecordType(inInner) && zng.IsRecordType(specInner) {
 				// 4. array/set of records
-				inInner := zng.AliasedType(innerType(inType))
-				specInner := zng.AliasedType(innerType(specType))
 				inner, err := s.cropRecordType(inInner.(*zng.TypeRecord), specInner.(*zng.TypeRecord))
 				if err != nil {
 					return nil, err
@@ -407,10 +407,10 @@ func (s *Shaper) orderRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, 
 	// with use, so starting with simpler algorithm for now.
 	//
 	for _, specCol := range spec.Columns {
-		specType := zng.AliasedType(specCol.Type)
 		if ind, ok := input.ColumnOfField(specCol.Name); ok {
 			inCol := input.Columns[ind]
 			inType := zng.AliasedType(inCol.Type)
+			specType := zng.AliasedType(specCol.Type)
 			if zng.IsRecordType(inType) && zng.IsRecordType(specType) {
 				if nested, err := s.orderRecordType(inType.(*zng.TypeRecord), specType.(*zng.TypeRecord)); err != nil {
 					return nil, err

--- a/expr/shaper.go
+++ b/expr/shaper.go
@@ -407,10 +407,12 @@ func (s *Shaper) orderRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, 
 	// with use, so starting with simpler algorithm for now.
 	//
 	for _, specCol := range spec.Columns {
+		specType := zng.AliasedType(specCol.Type)
 		if ind, ok := input.ColumnOfField(specCol.Name); ok {
 			inCol := input.Columns[ind]
-			if zng.IsRecordType(inCol.Type) && zng.IsRecordType(specCol.Type) {
-				if nested, err := s.orderRecordType(inCol.Type.(*zng.TypeRecord), specCol.Type.(*zng.TypeRecord)); err != nil {
+			inType := zng.AliasedType(inCol.Type)
+			if zng.IsRecordType(inType) && zng.IsRecordType(specType) {
+				if nested, err := s.orderRecordType(inType.(*zng.TypeRecord), specType.(*zng.TypeRecord)); err != nil {
 					return nil, err
 				} else {
 					cols = append(cols, zng.Column{specCol.Name, nested})
@@ -418,7 +420,9 @@ func (s *Shaper) orderRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, 
 				continue
 			}
 			if isCollectionType(inCol.Type) && isCollectionType(specCol.Type) && zng.IsRecordType(innerType(inCol.Type)) && zng.IsRecordType(innerType(specCol.Type)) {
-				if inner, err := s.orderRecordType(innerType(inCol.Type).(*zng.TypeRecord), innerType(specCol.Type).(*zng.TypeRecord)); err != nil {
+				inInner := zng.AliasedType(innerType(inCol.Type))
+				specInner := zng.AliasedType(innerType(specCol.Type))
+				if inner, err := s.orderRecordType(inInner.(*zng.TypeRecord), specInner.(*zng.TypeRecord)); err != nil {
 					return nil, err
 				} else {
 					var err error

--- a/expr/shaper.go
+++ b/expr/shaper.go
@@ -93,10 +93,10 @@ func createStepSet(in, out zng.Type) (step, error) {
 }
 
 func isCollectionType(t zng.Type) bool {
-	if _, ok := t.(*zng.TypeArray); ok {
-		return true
-	}
-	if _, ok := t.(*zng.TypeSet); ok {
+	switch t := t.(type) {
+	case *zng.TypeAlias:
+		return isCollectionType(t.Type)
+	case *zng.TypeArray, *zng.TypeSet:
 		return true
 	}
 	return false
@@ -104,6 +104,8 @@ func isCollectionType(t zng.Type) bool {
 
 func innerType(t zng.Type) zng.Type {
 	switch t := t.(type) {
+	case *zng.TypeAlias:
+		return innerType(t.Type)
 	case *zng.TypeArray:
 		return t.Type
 	case *zng.TypeSet:
@@ -506,7 +508,7 @@ func (s *Shaper) castRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 
 		if inCol.Type.ID() == specCol.Type.ID() {
 			// Field has same type in cast: output type unmodified.
-			cols = append(cols, inCol)
+			cols = append(cols, specCol)
 			continue
 		}
 		castType, err := s.castType(inCol.Type, specCol.Type)
@@ -523,12 +525,12 @@ func (c *Shaper) castType(inType, specType zng.Type) (zng.Type, error) {
 	switch {
 	case zng.IsRecordType(inType) && zng.IsRecordType(specType):
 		// Matching field is a record: recurse.
-		inRec := inType.(*zng.TypeRecord)
-		castRec := specType.(*zng.TypeRecord)
+		inRec := zng.AliasedType(inType).(*zng.TypeRecord)
+		castRec := zng.AliasedType(specType).(*zng.TypeRecord)
 		return c.castRecordType(inRec, castRec)
 	case zng.IsPrimitiveType(inType) && zng.IsPrimitiveType(specType):
 		// Matching field is a primitive: output type is cast type.
-		if LookupPrimitiveCaster(specType) == nil {
+		if LookupPrimitiveCaster(zng.AliasedType(specType)) == nil {
 			return nil, fmt.Errorf("cast to %s not implemented", specType)
 		}
 		return specType, nil
@@ -537,7 +539,7 @@ func (c *Shaper) castType(inType, specType zng.Type) (zng.Type, error) {
 		if err != nil {
 			return nil, err
 		}
-		if _, ok := specType.(*zng.TypeArray); ok {
+		if _, ok := zng.AliasedType(specType).(*zng.TypeArray); ok {
 			return c.zctx.LookupTypeArray(out), nil
 		}
 		return c.zctx.LookupTypeSet(out), nil

--- a/expr/shaper.go
+++ b/expr/shaper.go
@@ -447,11 +447,13 @@ func (s *Shaper) fillRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 	copy(cols, input.Columns)
 	for _, specCol := range spec.Columns {
 		if i, ok := input.ColumnOfField(specCol.Name); ok {
+			specType := zng.AliasedType(specCol.Type)
 			inCol := input.Columns[i]
+			inType := zng.AliasedType(inCol.Type)
 			// Field is present both in input and spec: recurse if
 			// both records, or select appropriate type if not.
-			if specRecType, ok := specCol.Type.(*zng.TypeRecord); ok {
-				if inRecType, ok := inCol.Type.(*zng.TypeRecord); ok {
+			if specRecType, ok := specType.(*zng.TypeRecord); ok {
+				if inRecType, ok := inType.(*zng.TypeRecord); ok {
 					filled, err := s.fillRecordType(inRecType, specRecType)
 					if err != nil {
 						return nil, err
@@ -462,21 +464,25 @@ func (s *Shaper) fillRecordType(input, spec *zng.TypeRecord) (*zng.TypeRecord, e
 				}
 				continue
 			}
-			if isCollectionType(inCol.Type) && isCollectionType(specCol.Type) && zng.IsRecordType(innerType(inCol.Type)) && zng.IsRecordType(innerType(specCol.Type)) {
-				if inner, err := s.fillRecordType(innerType(inCol.Type).(*zng.TypeRecord), innerType(specCol.Type).(*zng.TypeRecord)); err != nil {
-					return nil, err
-				} else {
-					var err error
-					var t zng.Type
-					if _, ok := inCol.Type.(*zng.TypeArray); ok {
-						t, err = s.zctx.LookupTypeArray(inner), nil
-					} else {
-						t, err = s.zctx.LookupTypeSet(inner), nil
-					}
-					if err != nil {
+			if isCollectionType(inType) && isCollectionType(specType) {
+				inInner := zng.AliasedType(innerType(inCol.Type))
+				specInner := zng.AliasedType(innerType(specCol.Type))
+				if zng.IsRecordType(inInner) && zng.IsRecordType(specInner) {
+					if inner, err := s.fillRecordType(inInner.(*zng.TypeRecord), specInner.(*zng.TypeRecord)); err != nil {
 						return nil, err
+					} else {
+						var err error
+						var t zng.Type
+						if _, ok := inCol.Type.(*zng.TypeArray); ok {
+							t, err = s.zctx.LookupTypeArray(inner), nil
+						} else {
+							t, err = s.zctx.LookupTypeSet(inner), nil
+						}
+						if err != nil {
+							return nil, err
+						}
+						cols[i] = zng.Column{specCol.Name, t}
 					}
-					cols[i] = zng.Column{specCol.Name, t}
 				}
 			}
 		} else {

--- a/expr/ztests/shape-alias.yaml
+++ b/expr/ztests/shape-alias.yaml
@@ -1,0 +1,8 @@
+zql: |
+  type port=int64; type prec={p: port, q: port}; type parr=[prec]; put .=shape({c: port, b:parr, a:prec})
+
+input: |
+  {a:{q:1, p: 2}, b:[{q:1}],c:3 }
+
+output: |
+  {c:3 (=port),b:[{p:null (port),q:1 (port)} (=0)] (=1),a:{p:2,q:1} (0)} (=2)

--- a/expr/ztests/shape-cast-alias.yaml
+++ b/expr/ztests/shape-cast-alias.yaml
@@ -8,5 +8,3 @@ input: |
 output: |
   {a:1 (=port),b:{p:1 (port)} (=0),c:[1 (port)] (=1)} (=2)
   {a:1,b:{p:1},c:[1]} (2)
-
-

--- a/expr/ztests/shape-cast-alias.yaml
+++ b/expr/ztests/shape-cast-alias.yaml
@@ -1,0 +1,12 @@
+zql: |
+  type port=int64; type prec={p: port}; type parr=[port]; put .=cast({a:port, b:prec, c: parr})
+
+input: |
+  {a: 1 (=myport), b:{p: 1 (myport)}, c:[1(myport)]}
+  {a: 1, b:{p: 1}, c:[1]}
+
+output: |
+  {a:1 (=port),b:{p:1 (port)} (=0),c:[1 (port)] (=1)} (=2)
+  {a:1,b:{p:1},c:[1]} (2)
+
+

--- a/expr/ztests/shape-crop-alias.yaml
+++ b/expr/ztests/shape-crop-alias.yaml
@@ -1,0 +1,10 @@
+zql: |
+  type port=int64; type prec={p: port}; type parr=[prec]; put .=crop({a:prec, b:parr})
+
+input: |
+  {a:{p:1 (port=(int64)), q: null (port)} (=0),b:[{p:1,q:2 (port)} (=1)] (=2),c:3 (port)} (=3)
+  {a:{p:1 (=myport),q:null (port)} (=4),b:[{p:1 (myport),q:2 (port)} (=5)] (=6),c:3 (port)} (=7)
+
+output: |
+  {a:{p:1 (=port)} (=0),b:[{p:1}]} (=1)
+  {a:{p:1 (=myport)} (=2),b:[{p:1} (2)] (=3)} (=4)

--- a/expr/ztests/shape-fill-alias.yaml
+++ b/expr/ztests/shape-fill-alias.yaml
@@ -1,0 +1,10 @@
+zql: |
+  type port=int64; type prec={p: port, q: port}; type parr=[prec]; put .=fill({a:prec, b:parr, c:port})
+
+input: |
+  {a: {q: 1}, b: [{p: 1}]}
+  {a: {q: 1 (=myport)}, b:[{p: 1 (myport)}]}
+
+output: |
+  {a:{q:1,p:null (port=(int64))} (=0),b:[{p:1,q:null (port)} (=1)] (=2),c:null (port)} (=3)
+  {a:{q:1 (=myport),p:null (port)} (=4),b:[{p:1 (myport),q:null (port)} (=5)] (=6),c:null (port)} (=7)

--- a/expr/ztests/shape-order-alias.yaml
+++ b/expr/ztests/shape-order-alias.yaml
@@ -1,0 +1,10 @@
+zql: |
+  type port=int64; type prec={p: port, q: port}; type parr=[prec]; put .=order({c: port, b:parr, a:prec})
+
+input: |
+  {a:{q:1 (port=(int64)), p: 2 (port)} (=0),b:[{q:1,p:2 (port)} (=1)] (=2),c:3 (port)} (=3)
+  {a:{q:1 (=myport), p: 2 (myport)} (=0),b:[{q:1 (myport),p:2 (myport)} (=1)] (=2),c:3 (myport)} (=3)
+
+output: |
+  {c:3 (=port),b:[{p:2 (port),q:1} (=0)] (=1),a:{p:2 (port),q:1 (port)} (=2)} (=3)
+  {c:3 (=myport),b:[{p:2 (myport),q:1 (myport)} (=4)] (=5),a:{p:2,q:1} (4)} (=6)

